2 :: Fixing relations, currencies now uses seq id

JonhyOliveira opened the pull request:
### **Schema**
- currencies: primary key changed from a string (`code`) to a sequential integer (`id`)
- other tables: updated foreign keys to reflect the change above.

### **Models**
Schema changes allows ActiveRecord associations to be less verbose, by leveraging Rails' defaults (`<Modelname>_id`).

### **Controllers/Views**
Some controllers, views and helpers that were making queries to the ActiveRecords that changed had to be updated.

JonhyOliveira committed:
Fixing relations, currencies now uses seq id

* **Schema**
    * currencies: primary key changed from a string (`code`) to a sequential integer (`id`)
    * other tables: updated foreign keys to reflect the change above.
* **Models**: Schema changes allows ActiveRecord associations to be less verbose, by leveraging Rails' defaults (`<Modelname>_id`).
* **Controllers/Views**: Some controllers, views and helpers that were making queries to the ActiveRecords that changed had to be updated.

JonhyOliveira commented:
Need to merge changes from other branches before applying these structural changes.

JonhyOliveira committed:
Merge branch 'main' into joao.oliveira/2-fixing-tables-primarykeys

JonhyOliveira committed:
Merge branch 'main' into joao.oliveira/2-fixing-tables-primarykeys

JonhyOliveira committed:
fixed some references to currency to use id

JonhyOliveira committed:
implemented feature

JonhyOliveira committed:
updated spec + 🤖rubocop

gdiasdasilva commented on rails_app/app/helpers/currencies_helper.rb:
```diff
@@ -33,7 +33,7 @@ def relativize_ex_rate(currency, round = true)
     # round if needed
     if round
-      ex_rate = ex_rate.round(user_signed_in? && current_user.favorite_currency != nil ? current_user.favorite_currency.decimal_digits : currency.decimal_digits)
+      ex_rate = ex_rate.round(user_signed_in? && current_user.favorite_currency != nil ? current_user.favorite_currency.decimal_digits : (currency.decimal_digits || 0))
```
I'd suggest something like this to make the code more readable. Inline conditionals with nested parentheses are not easy to read.

```ruby
ex_rate = if user_signed_in? && current_user.favorite_currency
  current_user.favorite_currency.decimal_digits
else
  currency.decimal_digits || 0
end

JonhyOliveira replied:
:+1: agree, didn't know that was a thing in ruby, but makes sense so it must be :smile:

gdiasdasilva commented on rails_app/app/helpers/currencies_helper.rb:
```diff
@@ -44,7 +44,7 @@ def relativize_record_ex_rate(record)
     # if the user has a favorite currency get the record for the same date and relativize
     if user_signed_in? && current_user.favorite_currency != nil
-      ex_rate = CurrencyRecord.find_by(code: current_user.favorite_currency.code,
+      ex_rate = CurrencyRecord.find_by(currency_id: current_user.favorite_currency.code,
```
Finding by ID but serving the currency code. Is this working as expected?

JonhyOliveira replied:
Probably not, fixing it.

gdiasdasilva commented on rails_app/db/migrate/20230420085710_change_currencies_pk.rb:
```diff
@@ -0,0 +1,8 @@
+  def change
+    execute "ALTER TABLE currencies DROP CONSTRAINT currencies_pkey CASCADE;"
+
+    add_column :currencies, :id, :primary_key
```
Is this really needed? Doesn't Rails assume the `id` as the PK by default?

JonhyOliveira replied:
I was not sure it would be able to assume that, since I hardcoded the PK drop. I also added `:primary_key` just to be sure

gdiasdasilva replied:
I think it assumes, I'm just not sure due to the overriding that were done before

gdiasdasilva commented on rails_app/db/migrate/20230420085950_change_relations_use_pk.rb:
```diff
@@ -0,0 +1,34 @@
+
+      t.remove_index([:follower_email, :followed_currency_code])
+
+      t.remove_foreign_key column: :follower_email if t.foreign_key_exists? column: :follower_email
+
+      t.remove_foreign_key column: :followed_currency_code if t.foreign_key_exists? column: :followed_currency_code
```
What's the reason to need those conditionals? I'd say at the time of the migration you can assume the FK exists. Otherwise we wouldn't be creating a migration to remove it.

JonhyOliveira replied:
:+1:

gdiasdasilva commented on rails_app/db/schema.rb:
```diff
@@ -29,17 +30,19 @@
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
-    t.index ["follower_email", "followed_currency_code"], name: "unique_on_follower_+_followed_currency", unique: true
+    t.bigint "currency_id"
```
Isn't this the same as the `id` column of this table that is automatically created when generating the `currencies` table?

JonhyOliveira replied:
This is to ensure db-side that there aren't duplicate records, e.g. multiple pairs (X, Y) of user X and currency Y. Also, it's useful when loading any currencies page where we need to check if a user follows a currency.

JonhyOliveira replied:
This is in the `currency_followings` table, it's the foreign key for the `currencies` table. Since github collapses lines, initially I did not notice that and had to go check the actual schema file. 
Did you not notice that too or am I missing something?

gdiasdasilva replied:
Oh, my bad. 😄 It was that, yes.

gdiasdasilva reviewed (changes requested):
Some of the previous features might not be working as expected but I'm not really sure. Left some comments!

JonhyOliveira committed:
Fixing readability

https://github.com/JonhyOliveira/H19-CurrencyManager/pull/3/files#r1178904013

JonhyOliveira committed:
Fixing potential bug

Referenced in https://github.com/JonhyOliveira/H19-CurrencyManager/pull/3/files#r1178905747

JonhyOliveira committed:
Migration is clearer now

Addressing https://github.com/JonhyOliveira/H19-CurrencyManager/pull/3/files#r1178909481

JonhyOliveira committed:
Fixing readability

Addressing https://github.com/JonhyOliveira/H19-CurrencyManager/pull/4/files#r1179239843

JonhyOliveira committed:
Making change_favorite easier to read

Addressing https://github.com/JonhyOliveira/H19-CurrencyManager/pull/4/files#r1179246067

JonhyOliveira committed:
View is less confusing

Adressing https://github.com/JonhyOliveira/H19-CurrencyManager/pull/4/files#r1179241163

gdiasdasilva commented on rails_app/app/helpers/currencies_helper.rb:
```diff
@@ -33,7 +33,14 @@ def relativize_ex_rate(currency, round = true)
+      end
+      
+      ex_rate = ex_rate.round(decimal_places)
+      puts "t"
```
What's this? 👀

JonhyOliveira committed:
Currency code query now uses `pluck`

Addressing https://github.com/JonhyOliveira/H19-CurrencyManager/pull/4#discussion_r1180224552

JonhyOliveira committed:
Removing a debug statement 🐞

Adressing https://github.com/JonhyOliveira/H19-CurrencyManager/pull/3#discussion_r1180216289

JonhyOliveira committed:
Merge pull request #4 from JonhyOliveira/joao.oliveira/3-edit-only-favorite-currency

3 :: View to edit only favourite currency
